rosbag2_cpp/src/rosbag2_cpp/writers/buffer_layer.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "rosbag2_cpp/writers/buffer_layer.hpp"
+
+#include "rosbag2_cpp/logging.hpp"
+
+namespace rosbag2_cpp
+{
+namespace writers
+{
+
+BufferLayer::BufferLayer(
+  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadWriteInterface> storage,
+  const uint64_t & max_buffer_size)
+: storage_(storage)
+{
+  // Set buffers
+  primary_buffer_ = std::shared_ptr<BagMessageBuffer>(
+    new BagMessageBuffer(max_buffer_size));
+  secondary_buffer_ =
+    std::shared_ptr<BagMessageBuffer>(
+    new BagMessageBuffer(max_buffer_size));
+
+  // Run consumer thread. We need this only when max cache size is not zero. Otherwise direct
+  // storage write call is used.
+  if (max_buffer_size > 0u) {
+    is_consumer_needed_ = true;
+    consumer_thread_ = std::thread(&BufferLayer::exec_consuming, this);
+  }
+}
+
+bool BufferLayer::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  // If cache size is set to zero, we directly call write
+  if (!is_consumer_needed_) {
+    storage_->write(msg);
+    return true;
+  }
+
+  // While pushing, we keep track of inserted and dropped messages as well
+  bool pushed = false;
+  {
+    std::lock_guard<std::mutex> buffer_lock(buffer_mutex_);
+    pushed = primary_buffer_->push(msg);
+  }
+
+  buffers_condition_var_.notify_one();
+
+  if (!pushed) {
+    elements_dropped_++;
+  }
+  return pushed;
+}
+
+void BufferLayer::exec_consuming()
+{
+  bool exit_flag = false;
+  while (true) {
```

Comment:
sorry but no! What about `while (!exit_flag)`? That would also ease the `return` logic at the end of the while loop.